pkg/repository/maintenance.go

```diff
@@ -0,0 +1,281 @@
+
+// MaintenanceConfig is the configuration for the repo maintenance job
+type MaintenanceConfig struct {
+	KeepLatestMaitenanceJobs int
+	CPURequest               string
+	MemRequest               string
+	CPULimit                 string
+	MemLimit                 string
+	LogLevelFlag             *logging.LevelFlag
+	FormatFlag               *logging.FormatFlag
+}
+
+func generateJobName(repo string) string {
+	millisecond := time.Now().UTC().UnixMilli() // millisecond
+
+	jobName := fmt.Sprintf("%s-maintain-job-%d", repo, millisecond)
+	if len(jobName) > 63 { // k8s job name length limit
+		jobName = fmt.Sprintf("repo-maintain-job-%d", millisecond)
+	}
+
+	return jobName
+}
+
+func getNodeSelectorFromVeleroServer(deployment *appsv1.Deployment) map[string]string {
+	return deployment.Spec.Template.Spec.NodeSelector
+}
+
+func getTolerationsFromVeleroServer(deployment *appsv1.Deployment) []v1.Toleration {
+	return deployment.Spec.Template.Spec.Tolerations
+}
+
+func getAffinityFromVeleroServer(deployment *appsv1.Deployment) *v1.Affinity {
+	return deployment.Spec.Template.Spec.Affinity
+}
+
+// getEnvVarsFromVeleroServer get the environment variables from the Velero server deployment
+func getEnvVarsFromVeleroServer(deployment *appsv1.Deployment) []v1.EnvVar {
+	for _, container := range deployment.Spec.Template.Spec.Containers {
+		// We only have one container in the Velero server deployment
+		return container.Env
+	}
+	return nil
+}
+
+// getVolumeMountsFromVeleroServer get the volume mounts from the Velero server deployment
+func getVolumeMountsFromVeleroServer(deployment *appsv1.Deployment) []v1.VolumeMount {
+	for _, container := range deployment.Spec.Template.Spec.Containers {
+		// We only have one container in the Velero server deployment
+		return container.VolumeMounts
+	}
+	return nil
+}
+
+// getVolumesFromVeleroServer get the volumes from the Velero server deployment
+func getVolumesFromVeleroServer(deployment *appsv1.Deployment) []v1.Volume {
+	return deployment.Spec.Template.Spec.Volumes
+}
+
+// getServiceAccountFromVeleroServer get the service account from the Velero server deployment
+func getServiceAccountFromVeleroServer(deployment *appsv1.Deployment) string {
+	return deployment.Spec.Template.Spec.ServiceAccountName
+}
+
+// getVeleroServerDeployment get the Velero server deployment
+func getVeleroServerDeployment(cli client.Client, namespace string) (*appsv1.Deployment, error) {
+	// Get the Velero server deployment
+	deployment := appsv1.Deployment{}
+	err := cli.Get(context.TODO(), types.NamespacedName{Name: "velero", Namespace: namespace}, &deployment)
+	if err != nil {
+		return nil, err
+	}
+	return &deployment, nil
+}
+
+func getVeleroServerImage(deployment *appsv1.Deployment) string {
+	// Get the image of the Velero server deployment
+	return deployment.Spec.Template.Spec.Containers[0].Image
+}
+
+func getResourceRequirements(cfg MaintenanceConfig) v1.ResourceRequirements {
```

Comment:
Let's move all the functions above to the util package, they are common functions and useful for future features.

Reply:
ok